kong/runloop/handler.lua

```diff
@@ -721,25 +728,117 @@ return {
 
       local trusted_ip = kong.ip.is_trusted(realip_remote_addr)
       if trusted_ip then
-        forwarded_proto = var.http_x_forwarded_proto or var.scheme
-        forwarded_host  = var.http_x_forwarded_host  or var.host
-        forwarded_port  = var.http_x_forwarded_port  or var.server_port
+        forwarded_proto = var.http_x_forwarded_proto or scheme
+        forwarded_host  = var.http_x_forwarded_host  or host
+        forwarded_port  = var.http_x_forwarded_port  or port
 
       else
-        forwarded_proto = var.scheme
-        forwarded_host  = var.host
-        forwarded_port  = var.server_port
+        forwarded_proto = scheme
+        forwarded_host  = host
+        forwarded_port  = port
       end
 
-      local protocols = route.protocols
-      if (protocols and
-          protocols.https and not protocols.http and forwarded_proto ~= "https")
+      local protocols = route and route.protocols
+      if (protocols and protocols.https and not protocols.http and
+          forwarded_proto ~= "https")
       then
         ngx.header["connection"] = "Upgrade"
         ngx.header["upgrade"]    = "TLS/1.2, HTTP/1.1"
         return kong.response.exit(426, { message = "Please use HTTPS protocol" })
       end
 
+      if not service then
+        -----------------------------------------------------------------------
+        -- Serviceless HTTP / HTTPS / HTTP2 route
+        -----------------------------------------------------------------------
+        local service_scheme
+        local service_host
+        local service_port
+
+        -- 1. try to find information from a request-line
+        if var.http2 == "" then
+          local request_line = var.request
+          if request_line then
+            local matches = re_match(request_line, [[\w+ (https?|wss?)://([^/?#\s]+)]], "ajos")
+            if matches then
+              local uri_scheme = lower(matches[1])
+              if uri_scheme == "https" or uri_scheme == "http" then
+                service_scheme = uri_scheme
+                service_host   = lower(matches[2])
+              elseif uri_scheme == "wss" then
```

Comment:
I wasn't 100% sure these paths were required, it was just an idea: did you verify what occurs if a websocket client uses a http proxy?

Reply:
@james-callahan, it needs testing. The devil is in details in this PR, so thank you for all the feedback so far. We need to get this right. Currently adding more tests to it.

Reply:
There is now more tests added. Lets figure out the websocket thing later as I think it needs in most cases `Connect`, and that is what we (or Nginx) don't support currently.